### PR TITLE
[aot] Switch Metal to SPIR-V codegen

### PR DESCRIPTION
**Pull request by PENGUINLIONG**
This PR is a pretty huge one:
1. Metal runtime is now written in Objective-C++ instead of wrapped C++.
2. The Metal backend is now driven by `GfxRuntime`. `MetalRuntime` has been removed entirely.
3. The Metal backend is now consuming artifacts from the SPIR-V codegen. SPIRV-Cross is linked to enable translation from SPIR-V to plain text MSL. The Metal codegen has been removed entirely.
4. The runtime C-API now supports Metal. Yay!
5. Adapted tests to be like other gfx archs. Removed sparse taichi extensions.

### CODE REVIEW

File: taichi/runtime/program_impls/metal/metal_program.h [hunk 1]
```diff
@@ -1,65 +1,102 @@
 #pragma once
```
Comment (PENGUINLIONG) [inline taichi/runtime/program_impls/metal/metal_program.h:1]: For this file I basically copied `vulkan_program.h`.

File: taichi/runtime/program_impls/metal/metal_program.cpp [hunk 2]
```diff
@@ -1,123 +1,95 @@
-#include <unordered_set>
```
Comment (PENGUINLIONG) [inline taichi/runtime/program_impls/metal/metal_program.cpp:1]: For this file I basically copied `vulkan_program.cpp`

File: taichi/codegen/spirv/spirv_codegen.cpp [hunk 3]
```diff
@@ -2156,7 +2156,7 @@ class TaskCodegen : public IRVisitor {
 
       // continue
       spirv::Value total_invocs = ir_->cast(
-          ir_->i32_type(),
+          ir_->u32_type(),
```
Comment (PENGUINLIONG) [inline taichi/codegen/spirv/spirv_codegen.cpp:2159]: ? how it passed CI

File: cmake/TaichiCore.cmake [hunk 4]
```diff
@@ -295,14 +300,12 @@ endif()
 if (TI_WITH_METAL)
     set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -DTI_WITH_METAL")
 
+    #if (NOT "${CMAKE_CXX_COMPILER_ID}" STREQUAL "AppleClang")
+    #    message(FATAL_ERROR "-- Metal backend must be compiled with AppleClang, but you are using ${CMAKE_CXX_COMPILER_ID}")
+    #endif()
```
Comment (PGZXB) [inline cmake/TaichiCore.cmake:305]: Remove these lines?

File: cmake/TaichiCore.cmake [hunk 5]
```diff
@@ -336,7 +339,7 @@ if (TI_WITH_OPENGL OR TI_WITH_VULKAN OR TI_WITH_DX11)
   target_link_libraries(${CORE_LIBRARY_NAME} PRIVATE gfx_runtime)
 endif()
```
Comment (PGZXB) [inline cmake/TaichiCore.cmake:340]: https://github.com/taichi-dev/taichi/pull/7093/files#diff-ddbed628f91a9ee8bb2f6af3cb6f4b793746b0cc8182bcd4513b6793e816f347R337

-> `if (TI_WITH_OPENGL OR TI_WITH_VULKAN OR TI_WITH_DX11 OR TI_WITH_METAL)` ?

**Review by PGZXB — APPROVED** [review]
LGTM!

File: taichi/codegen/spirv/spirv_codegen.cpp (hunk 3 above)
Comment (bobcao3) [reply taichi/codegen/spirv/spirv_codegen.cpp:2159]: The SPIR-V builder does absolutely no type checking at all...

File: taichi/rhi/metal/metal_device.h [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+namespace taichi::lang {
+
+namespace metal {
+
+struct MetalMemory;
+class MetalCommandList;
+class MetalStream;
+class MetalDevice;
+
+struct MetalMemory {
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.h:37]: when you have private members, just call this a class

File: taichi/rhi/metal/metal_device.mm [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+    MTLComputePipelineState_id mtl_compute_pipeline_state,
+    MetalWorkgroupSize workgroup_size)
+    : device_(&device), mtl_library_(mtl_library), mtl_function_(mtl_function),
+      mtl_compute_pipeline_state_(mtl_compute_pipeline_state),
+      workgroup_size_(workgroup_size) {}
+MetalPipeline::~MetalPipeline() { destroy(); }
+MetalPipeline *MetalPipeline::create(const MetalDevice &device,
+                                     const uint32_t *spv_data,
+                                     size_t spv_size) {
+  TI_ASSERT((size_t)spv_data % sizeof(uint32_t) == 0);
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.mm:36]: Don't use TI_ASSERT here (or any TI_xxx macros in fact), throw a exception and catch it from the Device API interface (& convert into return values, but that would need to wait for the API update for create_pipeline)

File: taichi/rhi/metal/metal_device.mm [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+MetalCommandList::~MetalCommandList() {}
+
+void MetalCommandList::bind_pipeline(Pipeline *p) noexcept {
+  TI_ASSERT(p != nullptr);
+  current_pipeline_ = (MetalPipeline *)p;
+}
+RhiResult MetalCommandList::bind_shader_resources(ShaderResourceSet *res,
+                                                  int set_index) noexcept {
+  TI_ASSERT(res != nullptr);
+  TI_ASSERT(set_index == 0);
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.mm:185]: same as above, don't use TI_ASSERT

File: taichi/rhi/metal/metal_device.mm [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+  current_shader_resource_set_ = (MetalShaderResourceSet *)res;
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.mm:186]: This command always happen after bind pipeline because bind pipeline will flush all bound resources

Comment (bobcao3) [reply taichi/rhi/metal/metal_device.mm:186]: So I don't think it's the best idea to do current_shader_resource_set_

File: taichi/rhi/metal/metal_device.mm [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+                    offset:resource.buffer.offset
+                   atIndex:resource.binding];
+        break;
+      }
+      default:
+        TI_ASSERT(false);
+      }
+    }
+
+    [encoder setComputePipelineState:mtl_compute_pipeline_state];
```
Comment (bobcao3) [inline taichi/rhi/metal/metal_device.mm:283]: Why not set pipeline state in the bind pipeline command?

**Pull request merged**